Milestone 1.2: Lesson translation dashboard.

DubeySandeep opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
This PR will allow an exploration owner to assign translators and then a translator can do translation work.

## Checklist
- [x] Backend controllers.
- [x] Frontend translation assigning work.
- [x] Fixing test errors.
- [x] Writing new tests.

DubeySandeep committed:
Initial changes

DubeySandeep committed:
Added translator controller

DubeySandeep committed:
Fixed test errors

DubeySandeep committed:
Made minor changes in handler

tjiang11 commented on core/controllers/translator.py:
```diff
@@ -0,0 +1,257 @@
+    """Handles requests from the translator for draft autosave."""
+
+    @acl_decorators.can_translate_exploration
+    def put(self, exploration_id):
```
Seeing how this code here seems exactly the same as the corresponding method in editor.py, could this perhaps be extracted out somewhere and then shared? I think I could see this being duplicated even more later on as new roles are added.

tjiang11 commented on core/controllers/translator.py:
```diff
@@ -0,0 +1,257 @@
+        self.render_json({})
+
+
+class ExplorationTranslationHandler(base.BaseHandler):
```
Same here. These methods look almost exactly the corresponding ones in ExplorationHandler in editor.py. Could we extract and reuse?

tjiang11 commented on core/domain/exp_services.py:
```diff
@@ -701,6 +701,7 @@ def apply_change_list(exploration_id, change_list):
         Exception: Any entries in the changelist are invalid.
     """
     exploration = get_exploration_by_id(exploration_id)
+    print change_list
```
Remove print statement.

DubeySandeep replied:
Done!

tjiang11 commented on core/domain/exp_services.py:
```diff
@@ -1125,7 +1126,7 @@ def publish_exploration_and_update_user_profiles(committer, exp_id):
 def update_exploration(
         committer_id, exploration_id, change_list, commit_message,
-        is_suggestion=False):
+        is_suggestion=False, is_translation_change=False):
```
Can we rename `is_translation_change` to `is_by_translator`?

DubeySandeep replied:
Done!

tjiang11 commented on core/domain/exp_services.py:
```diff
@@ -1141,6 +1142,8 @@ def update_exploration(
             feconf.COMMIT_MESSAGE_ACCEPTED_SUGGESTION_PREFIX.
         is_suggestion: bool. Whether the update is due to a suggestion being
             accepted.
+        is_translation_change: bool. Wheter the changes are made by a
```
Whether*

DubeySandeep replied:
Done!

tjiang11 commented on core/domain/exp_services.py:
```diff
@@ -1150,6 +1153,9 @@ def update_exploration(
             accepted suggestions.
     """
+    if is_translation_change and not is_translation_change_list(change_list):
+        raise Exception('Unable to save draft: Incorrect change_list.')
```
Can you change the exception message to something more like "Translator does not have permission to make some changes in the change list." It's a bit vague right now.

DubeySandeep replied:
Done!

tjiang11 commented on core/domain/exp_services.py:
```diff
@@ -1721,6 +1727,25 @@ def _is_suggestion_valid(thread_id, exploration_id):
 
 
+def is_translation_change_list(change_list):
+    """Checks whetehr the change list contains only the changes which are
```
whether*

DubeySandeep replied:
Done!

tjiang11 commented on core/domain/exp_services.py:
```diff
@@ -1721,6 +1727,25 @@ def _is_suggestion_valid(thread_id, exploration_id):
+            to be made to the ExplorationUserDataModel object.
+
+    Returns:
+        bool. Whether the change_list only the changes which are allowed for
```
Whether the change_list *contains* only...

DubeySandeep replied:
Done!

tjiang11 commented on core/domain/exp_services.py:
```diff
@@ -1924,7 +1950,12 @@ def create_or_update_draft(
             to be made to the ExplorationUserDataModel object.
         exp_version: int. The current version of the exploration.
         current_datetime: datetime.datetime. The current date and time.
+        is_translation_change: bool. Wheter the changes are made by a
```
Whether*

DubeySandeep replied:
Done!

tjiang11 commented on core/domain/exp_services.py:
```diff
@@ -1924,7 +1950,12 @@ def create_or_update_draft(
+            translator.
     """
+    if is_translation_change and not is_translation_change_list(change_list):
+        raise Exception('Unable to save draft: Incorrect change_list.')
```
Ditto

DubeySandeep replied:
Done!

tjiang11 commented on core/templates/dev/head/pages/exploration_editor/ExplorationDataService.js:
```diff
@@ -44,8 +44,15 @@ oppia.factory('ExplorationDataService', [
+    if (GLOBALS.can_edit) {
+      explorationDraftAutosaveUrl = (
+        '/createhandler/autosave_draft/' + explorationId);
+    } else {
```
I think it may be better to do `else if (GLOBALS.can_translate)` and then in the final else clause put a warning of some sort in alerts service. This just makes it more obvious for future changes of this sort.

DubeySandeep replied:
Done!

tjiang11 commented on core/templates/dev/head/pages/exploration_editor/settings_tab/settings_tab.html:
```diff
@@ -273,7 +282,7 @@ <h3>Roles</h3>
                   </select>
                   <span class="help-block">
-                    Note that managers also have the permissions of collaborators, and collaborators also have the permissions of viewers. Please note that assigning roles is irreversible (though you can always assign somebody to a higher role).
+                    Note that managers also have the permissions of collaborators, collaborators also have the permissions of translators, and translators also have permissions of viewers. Please note that assigning roles is irreversible (though you can always assign somebody to a higher role).
```
...translators also have *the* permissions of viewers...

DubeySandeep replied:
Done!

tjiang11 reviewed (changes requested):
Thanks Sandeep!

Issues: 

Translators can try to add a new response or hint, but it results in a backend error.

Translators should not be allowed to edit or add a solution. This results in a backend error when attempted.

tjiang11 commented:
I would go ahead and write the tests

DubeySandeep committed:
Added review changes (1)

DubeySandeep commented:
@tjiang. Made review changes and no translator will not be able to add answers/hints (I had done a small mistake in EditabilityServices.js which is fixed now).

Regarding:
>Seeing how this code here seems exactly the same as the corresponding method in editor.py, could this perhaps be extracted out somewhere and then shared? I think I could see this being duplicated even more later on as new roles are added.

I'm going to bed now I'll let you know If I'll come up with an idea to remove duplicity.

DubeySandeep committed:
Added Initials for e2e tests and fixed lint errors

tjiang11 reviewed (changes requested):
You forgot to rename some variable names so it's inconsistent and causing backend errors. (is_translation_change)

I think this looks strange. Can you at least put something like "No hints have been created yet." for translators or similar?

![image](https://user-images.githubusercontent.com/11807091/41355869-131660d0-6ee8-11e8-9608-120c8935e97c.png)

(Same with solutions)

As for removing the duplicate code, I'm not sure. I think it may be acceptable to extract it into exp_services, what do you think? Let's try not to stall on this since time is of the essence.

DubeySandeep committed:
Added e2e to see an owner can assign translator

DubeySandeep committed:
Added backend and frontend tests

seanlip commented:
Hi @DubeySandeep -- FYI, there are lint errors?

DubeySandeep committed:
Added backend test for translator handlers

DubeySandeep commented:
@tjiang11 
Regarding:
> I think this looks strange. Can you at least put something like "No hints have been created yet." for translators or similar?

I found that it's the current behaviour of oppia's editor page i.e, when a logged in user (not the editor) goes to exp editor page this is what they see. I have removed the "Hint" where hint == 0 and not editor.

Also, you can take a final review as I have added the backend tests.

DubeySandeep committed:
Added minor nit changes

tjiang11 commented on core/controllers/translator_test.py:
```diff
@@ -0,0 +1,225 @@
+            response['states'][state_name]['content_ids_to_audio_translations'],
+            self.CONTENT_IDS_TO_AUDIO_TRANSLATION)
+
+    def test_transator_cannot_save_invalid_change_list(self):
```
translator*

DubeySandeep replied:
Done!

tjiang11 commented on core/controllers/translator_test.py:
```diff
@@ -0,0 +1,225 @@
+                'commit_message': 'Translated first state content',
+                'version': 1
+            }, self.csrf_token)
+        # Checking the response to have audio tarnslations.
```
translations*

DubeySandeep replied:
Done!

tjiang11 commented on core/controllers/translator_test.py:
```diff
@@ -0,0 +1,225 @@
+                    'property_name': 'objective',
+                    'new_value': 'the objective',
+                }],
+                'commit_message': 'Chnaged exp objective',
```
Changed*

DubeySandeep replied:
Done!

tjiang11 reviewed (changes requested):
Hi Sandeep, in getting rid of the edit response/hint buttons for translators, you got rid of the save draft button for translators. (Test your code)

DubeySandeep committed:
Added review changes

DubeySandeep commented:
@tjiang11, I've added the review changes PTAL!

tjiang11 commented:
LGTM, thanks! Can merge after tests pass
